src/cstruct_ext.ml

```diff
@@ -0,0 +1,20 @@
+(* [or_empty cs] is [cs] unless [Cstruct.is_empty cs] where it is [Cstruct.empty].
+   An empty cstruct will keep a live reference to its underlying buffer. It is
+   preferable to keep a reference live to [Cstruct.empty.buffer] than
+   [cs.buffer] if [cs] is empty. *)
+let[@ocaml.inline always] or_empty cs =
+  if Cstruct.is_empty cs then Cstruct.empty else cs
+
+(* [Cstruct.sub cs 0 cs.len] will result in an empty cstruct **that keeps a
+   live reference to [cs.buffer]**!! *)
+let sub cs off len = or_empty (Cstruct.sub cs off len)
```

Comment:
this hides potential errors, no? if I do a "Cstruct.sub Cstruct.empty 0 10", now I get the Cstruct.empty back -- previously I got an exception (not sure whether this is relevant or not).